File: Python/insertionSort.py
def insertion_sort(arr):
    
    # Traverse through 1 to the length of the array
    for i in range(1, len(arr)):
        key = arr[i]  # The current element to be inserted
        j = i - 1  # The last index of the sorted portion
        
        # Move elements of arr[0..i-1], that are greater than key,
        # to one position ahead of their current position
        while j >= 0 and key < arr[j]:
            arr[j + 1] = arr[j]  # Shift the element to the right
            j -= 1  # Move to the left in the sorted portion
        
        arr[j + 1] = key  # Place the key in its correct position

    return arr  # Return the sorted array

# Example usage
if __name__ == "__main__":
    # Sample array to sort
    sample_array = [12, 11, 13, 5, 6]
    print("Original array:", sample_array)
    
    # Sorting the array
    sorted_array = insertion_sort(sample_array)
    print("Sorted array:", sorted_array)

File: JavaScript/insertionSort.js
/**
 * Function to perform Insertion Sort on an array.
 * @param {number[]} arr - The array to be sorted.
 * @returns {number[]} - The sorted array.
 */
function insertionSort(arr) {
    let n = arr.length; // Get the length of the array

    // Iterate through the array starting from the second element
    for (let i = 1; i < n; i++) {
        let currentElement = arr[i]; // The current element to be positioned
        let j = i - 1; // Start comparing with the last sorted element

        // Shift elements of arr[0..i-1], that are greater than currentElement,
        // to one position ahead of their current position
        while (j >= 0 && arr[j] > currentElement) {
            arr[j + 1] = arr[j]; // Shift the larger element to the right
            j--; // Move to the next element on the left
        }

        // Place the current element at its correct position
        arr[j + 1] = currentElement;
    }

    return arr; // Return the sorted array
}

// Example usage
const arrayToSort = [64, 34, 25, 12, 22, 11, 90];
const sortedArray = insertionSort(arrayToSort);
console.log("Sorted Array:", sortedArray);
